File: assignments/elixir/custom-set/custom-set_test.exs
Code.load_file("custom_set.exs")
ExUnit.start

defmodule CustomSetTest do
  use ExUnit.Case
  doctest CustomSet

  # Tip: to make your set work with the Set module be sure that your set value
  # is a tuple with as first element `CustomSet`. A record with tag `CustomSet`
  # will meet this requirement.

  test "delete" do
    assert Set.equal?(Set.delete(CustomSet.new([3,2,1]), 2), CustomSet.new([1,3]))
    assert Set.equal?(Set.delete(CustomSet.new([3,2,1]), 4), CustomSet.new([1,2,3]))
    assert Set.equal?(Set.delete(CustomSet.new([3,2,1]), 2.0), CustomSet.new([1,2,3]))
  end

  test "difference" do
    assert Set.equal?(
      Set.difference(CustomSet.new([1,2,3]), CustomSet.new([2,4])),
      CustomSet.new([1,3]))
    assert Set.equal?(
      Set.difference(CustomSet.new([1,2.0,3]), CustomSet.new([2,4])),
      CustomSet.new([1,2.0,3]))
  end

  test "disjoint?" do
    assert Set.disjoint?(CustomSet.new([1,2]), CustomSet.new([3,4]))
    refute Set.disjoint?(CustomSet.new([1,2]), CustomSet.new([2,3]))
    assert Set.disjoint?(CustomSet.new([1.0,2.0]), CustomSet.new([2,3]))
    assert Set.disjoint?(CustomSet.new, CustomSet.new)
  end

  test "empty" do
    assert Set.equal?(Set.empty(CustomSet.new([1,2])), CustomSet.new)
    assert Set.equal?(Set.empty(CustomSet.new), CustomSet.new)
  end

  test "intersection" do
    assert Set.equal?(
      Set.intersection(CustomSet.new([:a, :b, :c]), CustomSet.new([:a, :c, :d])),
      CustomSet.new([:a, :c]))
    assert Set.equal?(
      Set.intersection(CustomSet.new([1, 2, 3]), CustomSet.new([1.0, 2.0, 3])),
      CustomSet.new([3]))
  end

  test "member?" do
    assert Set.member?(CustomSet.new([1,2,3]), 2)
    assert Set.member?(CustomSet.new(1..3), 2)
    refute Set.member?(CustomSet.new(1..3), 2.0)
    refute Set.member?(CustomSet.new(1..3), 4)
  end

  test "put" do
    assert Set.equal?(
      Set.put(CustomSet.new([1,2,4]), 3),
      CustomSet.new([1,2,3,4]))
    assert Set.equal?(
      Set.put(CustomSet.new([1,2,3]), 3),
      CustomSet.new([1,2,3]))
    assert Set.equal?(
      Set.put(CustomSet.new([1,2,3]), 3.0),
      CustomSet.new([1,2,3,3.0]))
  end

  test "size" do
    assert Set.size(CustomSet.new) == 0
    assert Set.size(CustomSet.new([1,2,3])) == 3
    assert Set.size(CustomSet.new([1,2,3,2])) == 3
  end

  test "subset?" do
    assert Set.subset?(CustomSet.new([1,2,3]), CustomSet.new([1,2,3]))
    assert Set.subset?(CustomSet.new([1,2,3]), CustomSet.new([4,1,2,3]))
    refute Set.subset?(CustomSet.new([1,2,3]), CustomSet.new([4,1,3]))
    refute Set.subset?(CustomSet.new([1,2,3.0]), CustomSet.new([1,2,3,4]))
    assert Set.subset?(CustomSet.new, CustomSet.new([4,1,3]))
    assert Set.subset?(CustomSet.new, CustomSet.new)
  end

  test "to_list" do
    assert Enum.sort(Set.to_list(CustomSet.new)) == []
    assert Enum.sort(Set.to_list(CustomSet.new([3,1,2]))) == [1,2,3]
    assert Enum.sort(Set.to_list(CustomSet.new([3,1,2,1]))) == [1,2,3]
  end

  test "union" do
    assert Set.equal?(
      CustomSet.union(CustomSet.new([1,3]), CustomSet.new([2])),
      CustomSet.new([3,2,1]))
    assert Set.equal?(
      CustomSet.union(CustomSet.new([1,3]), CustomSet.new([2,3.0])),
      CustomSet.new([3.0,3,2,1]))
    assert Set.equal?(
      CustomSet.union(CustomSet.new([1,3]), CustomSet.new),
      CustomSet.new([3,1]))
    assert Set.equal?(
      CustomSet.union(CustomSet.new, CustomSet.new([2])),
      CustomSet.new([2]))
    assert Set.equal?(
      CustomSet.union(CustomSet.new, CustomSet.new),
      CustomSet.new([]))
  end

  test "inspect" do
    assert inspect(CustomSet.new) == "#<CustomSet []>"
    assert inspect(CustomSet.new([1,3,2])) == "#<CustomSet [1, 2, 3]>"
    # Weird ordering due to how Elixir/Erlang considers binaries, numbers and
    # atoms to be ordered.
    assert inspect(CustomSet.new(["A",:b,?C])) == "#<CustomSet [67, :b, \"A\"]>"
  end
end

File: assignments/ruby/custom-set/custom_set_test.rb
require 'minitest/autorun'
require_relative 'custom_set'

class CustomSetTest < MiniTest::Unit::TestCase

  def test_equal
    assert_equal CustomSet.new([1, 3]), CustomSet.new([3, 1])
  end

  def test_delete
    assert_equal CustomSet.new([1,3]), CustomSet.new([3,2,1]).delete(2)
    assert_equal CustomSet.new([1,2,3]), CustomSet.new([3,2,1]).delete(4)
    assert_equal CustomSet.new([1,2,3]), CustomSet.new([3,2,1]).delete(2.0)
  end

  def test_difference
    assert_equal CustomSet.new([1,3]),
      CustomSet.new([1,2,3]).difference(CustomSet.new([2,4]))

    assert_equal CustomSet.new([1,2.0,3]),
      CustomSet.new([1,2.0,3]).difference(CustomSet.new([2,4]))
  end

  def test_disjoint?
    assert CustomSet.new([1,2]).disjoint?(CustomSet.new([3,4]))
    refute CustomSet.new([1,2]).disjoint?(CustomSet.new([2,3]))
    assert CustomSet.new([1.0,2.0]).disjoint?(CustomSet.new([2,3]))
    assert CustomSet.new.disjoint?(CustomSet.new)
  end

  def test_empty
    assert_equal CustomSet.new, CustomSet.new([1,2]).empty
    assert_equal CustomSet.new, CustomSet.new.empty
  end

  def test_intersection
    assert_equal CustomSet.new([:a, :c]),
      CustomSet.new([:a, :b, :c]).intersection(CustomSet.new([:a, :c, :d]))

    assert_equal CustomSet.new([3]),
      CustomSet.new([1, 2, 3]).intersection(CustomSet.new([1.0, 2.0, 3]))
  end

  def test_member?
    assert CustomSet.new([1,2,3]).member?(2)
    assert CustomSet.new(1..3).member?(2)
    refute CustomSet.new(1..3).member?(2.0)
    refute CustomSet.new(1..3).member?(4)
  end

  def test_put
    assert_equal CustomSet.new([1,2,3,4]),
      CustomSet.new([1,2,4]).put(3)

    assert_equal CustomSet.new([1,2,3]),
      CustomSet.new([1,2,3]).put(3)

    assert_equal CustomSet.new([1,2,3,3.0]),
      CustomSet.new([1,2,3]).put(3.0)
  end

  def test_size
    assert_equal 0, CustomSet.new.size
    assert_equal 3, CustomSet.new([1,2,3]).size
    assert_equal 3, CustomSet.new([1,2,3,2]).size
  end

  def test_subset?
    assert CustomSet.new([1,2,3]).subset?(CustomSet.new([1,2,3]))
    assert CustomSet.new([4,1,2,3]).subset?(CustomSet.new([1,2,3]))
    refute CustomSet.new([4,1,3]).subset?(CustomSet.new([1,2,3]))
    refute CustomSet.new([1,2,3,4]).subset?(CustomSet.new([1,2,3.0]))
    assert CustomSet.new([4,1,3]).subset?(CustomSet.new)
    assert CustomSet.new.subset?(CustomSet.new)
  end

  def test_to_list
    assert_equal [], CustomSet.new.to_list.sort
    assert_equal [1,2,3], CustomSet.new([3,1,2]).to_list.sort
    assert_equal [1,2,3], CustomSet.new([3,1,2,1]).to_list.sort
  end

  def test_union
    assert_equal CustomSet.new([3,2,1]),
      CustomSet.new([1,3]).union(CustomSet.new([2]))
    assert_equal CustomSet.new([3.0,3,2,1]),
      CustomSet.new([1,3]).union(CustomSet.new([2,3.0]))
    assert_equal CustomSet.new([3,1]),
      CustomSet.new([1,3]).union(CustomSet.new)
    assert_equal CustomSet.new([2]),
      CustomSet.new([2]).union(CustomSet.new)
    assert_equal CustomSet.new([]),
      CustomSet.new.union(CustomSet.new)
  end

end
